### PR TITLE
[SPARK-47491][CORE] Add `slf4j-api` jar to the class path first before the others of `jars` directory

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to add `slf4j-api` jar to the class path before the others of `jars` directory.

This will re-enable `driver log links` test in `YarnClusterSuite`, too.

### Why are the changes needed?

**BEFORE**
- The test fails due to the following in GitHub Action CI environment.

```
24/03/20 18:32:54.385 launcher-proc-24 INFO YarnClusterDriver: SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
24/03/20 18:32:54.385 launcher-proc-24 INFO YarnClusterDriver: SLF4J: Defaulting to no-operation (NOP) logger implementation
24/03/20 18:32:54.385 launcher-proc-24 INFO YarnClusterDriver: SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

```
[info] - running Spark in yarn-cluster mode displays driver log links *** FAILED *** (14 seconds, 21 milliseconds)
[info]   java.io.FileNotFoundException: /home/runner/work/spark/spark/target/tmp/spark-6660465a-18fd-41d7-be7c-a8f63b3c83ef/logs (No such file or directory)
```

**AFTER**
- https://github.com/dongjoon-hyun/spark/actions/runs/8448089836/job/23139537396
```
[info] YarnClusterSuite:
[info] - run Spark in yarn-client mode (16 seconds, 37 milliseconds)
[info] - run Spark in yarn-cluster mode (15 seconds, 28 milliseconds)
[info] - run Spark in yarn-client mode with unmanaged am (12 seconds, 23 milliseconds)
[info] - run Spark in yarn-client mode with different configurations, ensuring redaction (14 seconds, 34 milliseconds)
[info] - run Spark in yarn-cluster mode with different configurations, ensuring redaction (15 seconds, 24 milliseconds)
[info] - yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630) (13 seconds, 23 milliseconds)
[info] - SPARK-35672: run Spark in yarn-client mode with additional jar using URI scheme 'local' (15 seconds, 26 milliseconds)
[info] - SPARK-35672: run Spark in yarn-cluster mode with additional jar using URI scheme 'local' (15 seconds, 22 milliseconds)
[info] - SPARK-35672: run Spark in yarn-client mode with additional jar using URI scheme 'local' and gateway-replacement path (14 seconds, 22 milliseconds)
[info] - SPARK-35672: run Spark in yarn-cluster mode with additional jar using URI scheme 'local' and gateway-replacement path (15 seconds, 21 milliseconds)
[info] - SPARK-35672: run Spark in yarn-cluster mode with additional jar using URI scheme 'local' and gateway-replacement path containing an environment variable (15 seconds, 28 milliseconds)
[info] - SPARK-35672: run Spark in yarn-client mode with additional jar using URI scheme 'file' (14 seconds, 28 milliseconds)
[info] - SPARK-35672: run Spark in yarn-cluster mode with additional jar using URI scheme 'file' (15 seconds, 22 milliseconds)
[info] - run Spark in yarn-cluster mode unsuccessfully (12 seconds, 23 milliseconds)
[info] - run Spark in yarn-cluster mode failure after sc initialized (23 seconds, 25 milliseconds)
[info] - run Python application in yarn-client mode (16 seconds, 30 milliseconds)
[info] - run Python application in yarn-cluster mode (17 seconds, 27 milliseconds)
[info] - run Python application in yarn-cluster mode using spark.yarn.appMasterEnv to override local envvar (18 seconds, 26 milliseconds)
[info] - user class path first in client mode (15 seconds, 21 milliseconds)
[info] - user class path first in cluster mode (15 seconds, 22 milliseconds)
[info] - monitor app using launcher library (8 seconds, 501 milliseconds)
[info] - running Spark in yarn-cluster mode displays driver log links (16 seconds, 70 milliseconds)
[info] - timeout to get SparkContext in cluster mode triggers failure (16 seconds, 19 milliseconds)
[info] - executor env overwrite AM env in client mode (14 seconds, 23 milliseconds)
[info] - executor env overwrite AM env in cluster mode (14 seconds, 20 milliseconds)
[info] - SPARK-34472: ivySettings file with no scheme or file:// scheme should be localized on driver in cluster mode (29 seconds, 47 milliseconds)
[info] - SPARK-34472: ivySettings file with no scheme or file:// scheme should retain user provided path in client mode (28 seconds, 37 milliseconds)
[info] - SPARK-34472: ivySettings file with non-file:// schemes should throw an error (4 seconds, 37 milliseconds)
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs with re-enabled test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
YARN tests passed.

Screenshot 2024-03-27 at 01 01 31

**Comment by dongjoon-hyun** [comment]
Could you review this PR, @HyukjinKwon ? This will recover the Maven Daily CI failure too.

File: launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java [hunk 1]
```diff
@@ -193,6 +193,12 @@ List<String> buildClassPath(String appClassPath) throws IOException {
     boolean isTestingSql = "1".equals(getenv("SPARK_SQL_TESTING"));
     String jarsDir = findJarsDir(getSparkHome(), getScalaVersion(), !isTesting && !isTestingSql);
     if (jarsDir != null) {
+      // Place slf4j-api-* jar first to be robust
```
Comment (HyukjinKwon) [inline launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java:196]: For my own understanding, is this the test specific issue, or actual bug?

**Review by HyukjinKwon — APPROVED** [review]

File: launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java (hunk 1 above)
Comment (dongjoon-hyun) [reply launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java:196]: Ya, it seems to be a real robustness issue. It randomly happens based on the class loading sequence.

**Comment by dongjoon-hyun** [comment]
Thank you for review and approval, @HyukjinKwon . 

After recovering this YARN test and daily Maven CIs via this PR, we can revisit this area later while keeping CIs green.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

**Comment by dongjoon-hyun** [comment]
For the record, Maven CI is recovered with this patch.
- https://github.com/apache/spark/actions/runs/8452197772 (`master` Java 17 Maven)

**Comment by LuciferYang** [comment]
late LGTM

**Comment by LuciferYang** [comment]
@dongjoon-hyun 

It's a bit magical, I found that in the latest test, `Run / Build modules using Maven: repl,sql#hive-thriftserver` also turned green, but before yesterday it was still red. May I ask if this is an additional benefit for this PR?

-  14 hours ago: https://github.com/apache/spark/actions/runs/8452197772/job/23152049324
- 2 days ago: https://github.com/apache/spark/actions/runs/8436403960/job/23138388265

**Comment by dongjoon-hyun** [comment]
Yes, @LuciferYang , as I mentioned in the above comment, I was also investigating them too because they failed at the log checking stage at that time due to `SLF4J: Defaulting to no-operation (NOP) logger implementation`.
- https://github.com/apache/spark/pull/45618#issuecomment-2022177524

Before this PR, all Spark log output could disappear. In our CI, `Spark Driver` log and `Spark Thrift Server` log was missing. In the same way, all Spark JVMs like `Spark Master` and `Spark Worker` could be affected in the same way randomly before this PR.
> May I ask if this is an additional benefit for this pr? Thanks ~

**Comment by LuciferYang** [comment]
Thanks  for your explanation @dongjoon-hyun ~

**Pull request closed without merging**